### PR TITLE
Js: esnext

### PR DESCRIPTION
**Pull request by studentmain**
Update js grammar to recent standard

### CODE REVIEW

**Comment by studentmain** [comment]
As CI not works in JS grammar, a manually test may need.

Known problem, help wanted:
- Can't parse `for of`. Works with IDEA plugin, fail in grun (p("of") fail)
  - Tried `Of: "of";`, but it still recognized as Identifier.
- Nested template string won't work: \`outer${\`inner\`}\`

File: javascript/JavaScriptLexer.g4 [hunk 1]
```diff
@@ -33,6 +34,7 @@ channels { ERROR }
 
 options { superClass=JavaScriptBaseLexer; }
 
+HashBangLine:                   '#!' ~[\r\n\u2028\u2029]*; // only allowed at start
```
Comment (KvanTTT) [inline javascript/JavaScriptLexer.g4]: Have you seen how it is handled in PHP grammar? https://github.com/antlr/grammars-v4/blob/master/php/PhpLexer.g4#L46

File: javascript/README.md [hunk 2]
```diff
@@ -90,6 +90,19 @@ Grammar supports the following list of ECMAScript 6 features taken from
 
 See [examples](examples) directory with test data files.
 
+### ES6 to ES2020
```
Comment (KvanTTT) [inline javascript/README.md]: Please always use two line breaks after headers.

File: javascript/README.md [hunk 3]
```diff
@@ -105,6 +118,17 @@ Also this grammar supports outdated syntax such as
   * Cleared & optimized
   * Universal code actions & predicates
   * Support of some outdated syntax (Html comment, CData)
+* Student Main (2019):
+  * Update to ES2020
+
+## Running fuzz test
```
Comment (KvanTTT) [inline javascript/README.md]: Insert line break.

File: javascript/JavaScriptParser.g4 [hunk 4]
```diff
@@ -184,7 +188,7 @@ tryStatement
     ;
 
 catchProduction
-    : Catch '(' Identifier ')' block
+    : Catch ('(' (Identifier|arrayLiteral|objectLiteral)? ')')? block
```
Comment (KvanTTT) [inline javascript/JavaScriptParser.g4]: Please, always use space before and after `|` character.

File: javascript/README.md [hunk 5]
```diff
@@ -90,6 +90,21 @@ Grammar supports the following list of ECMAScript 6 features taken from
 
 See [examples](examples) directory with test data files.
 
+
```
Comment (KvanTTT) [inline javascript/README.md]: But I meant to insert a line break after the header but not before :) Also, see [MD022](https://github.com/DavidAnson/markdownlint/blob/master/doc/Rules.md#md022).

Comment (studentmain) [reply javascript/README.md]: My mistake. Fixed.

**Comment by KvanTTT** [comment]
> Known problem, help wanted:

Maybe extract them to a new issue? We can fix them later.

**Comment by studentmain** [comment]
> Maybe extract them to a new issue?

Sounds good. I'll list not implemented feature in a new issue here later.

But about `for of`, I'm not sure where the problem is. Post it in another issue?

**Comment by studentmain** [comment]
Finished, ready to merge.

**Comment by KvanTTT** [comment]
> But about for of, I'm not sure where the problem is. Post it in another issue?

Yes, I think so.

**Pull request merged**